crates/ruff_db/src/diagnostic.rs

```diff
@@ -1,16 +1,146 @@
+use std::borrow::Cow;
+use std::fmt::Formatter;
+
+use thiserror::Error;
+
+use ruff_python_parser::ParseError;
+use ruff_text_size::TextRange;
+
 use crate::{
     files::File,
     source::{line_index, source_text},
     Db,
 };
-use ruff_python_parser::ParseError;
-use ruff_text_size::TextRange;
-use std::borrow::Cow;
+
+/// A string identifier for a lint rule.
+///
+/// This string is used in command line and configuration interfaces. The name should always
+/// be in kebab case, e.g. `no-foo` (all lower case).
+///
+/// Rules use kebab case, e.g. `no-foo`.
+#[derive(Debug, Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct LintName(&'static str);
+
+impl LintName {
+    pub const fn of(name: &'static str) -> Self {
+        Self(name)
+    }
+
+    pub const fn as_str(&self) -> &'static str {
+        self.0
+    }
+}
+
+impl std::ops::Deref for LintName {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl std::fmt::Display for LintName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.0)
+    }
+}
+
+impl PartialEq<str> for LintName {
+    fn eq(&self, other: &str) -> bool {
+        self.0 == other
+    }
+}
+
+impl PartialEq<&str> for LintName {
+    fn eq(&self, other: &&str) -> bool {
+        self.0 == *other
+    }
+}
+
+/// Uniquely identifies the kind of a diagnostic.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum DiagnosticId {
+    /// Some I/O operation failed
+    Io,
+
+    /// Some code contains a syntax error
+    InvalidSyntax,
+
+    /// A lint violation.
+    ///
+    /// Lints can be suppressed and some lints can be enabled or disabled in the configuration.
+    Lint(LintName),
+
+    /// A revealed type: Created by `reveal_type(expression)`.
+    RevealedType,
+}
+
+impl DiagnosticId {
+    /// Creates a new `DiagnosticId` for a lint with the given name.
+    pub const fn lint(name: &'static str) -> Self {
+        Self::Lint(LintName::of(name))
+    }
+
+    /// Returns `true` if this `DiagnosticId` represents a lint.
+    pub fn is_lint(&self) -> bool {
+        matches!(self, DiagnosticId::Lint(_))
+    }
+
+    /// Returns `true` if this `DiagnosticId` represents a lint with the given name.
+    pub fn is_lint_named(&self, name: &str) -> bool {
+        matches!(self, DiagnosticId::Lint(self_name) if self_name == name)
+    }
+
+    pub fn matches(&self, name: &str) -> bool {
+        match self.as_str() {
+            Ok(id) => id == name,
+            Err(DiagnosticAsStrError::Category { category, name }) => name
+                .strip_prefix(category)
+                .and_then(|prefix| prefix.strip_prefix(":"))
+                .is_some_and(|rest| rest == name),
+        }
+    }
+
+    pub fn as_str(&self) -> Result<&str, DiagnosticAsStrError> {
```

Comment:
Would it make more sense to just return a `DiagnosticAsStr` type that implements `Display`? So, basically, a `DiagnosticAsStrError` like you have, but for all the variants. And in this case, you'd probably rename it to `display()` or some such.

Reply:
Another pattern I like for things like this is to bite the bullet on the alloc, but do it at construction time. Then you can have an `as_str()` method that unconditionally returns `&str`.

Reply:
I introduced it to avoid repeating the variant to name matching code for the `matches` method. I think the easiest would be to just call `to_string()` in matches which I'm open to do when we have other use cases for it. 

Another solution is to have a macro that generates a `name` method for it. 

Overall, I don't have a very good understanding yet of how this type will be used, and I suggest we change it as we see fit.